### PR TITLE
Faucet page ICR for coin info + consolidation of client-side queries

### PR DESCRIPTION
**Pull request by bogberry**
- Utilizes incremental static regeneration to minimize the number of request made on client side on faucet page
- Use `useQueries` on page component to simplify balance querying logic and consolidate queries into one place
- CSS layout changes so mint cards stack on top of each other at narrow widths
- Switch to reading faucet address from environment variable instead of a hardcoded value

### CODE REVIEW

**Comment by bogberry** [comment]
@alnoki where is the faucet published on testnet? Build fails because the faucet cannot be found.

**Comment by alnoki** [comment]
> @alnoki where is the faucet published on testnet? Build fails because the faucet cannot be found.

@bogberry https://github.com/econia-labs/econia/tree/1ac68bb5/src/move/econia is now live on the single-signer Econia testnet account: https://explorer.aptoslabs.com/account/0x40b119411c6a975fca28f1ba5800a8a418bba1e16a3f13b1de92f731e023d135/modules/code/assets?network=testnet

Note the cautionary badge on our docs homepage about testnet moving to a multisig soon (after audit on new event code and deployment to mainnet we can switch testnet to be a multisig to ensure same testing conditions as mainnet): https://econia.dev/#account-addresses

File: src/typescript/frontend/src/pages/faucet/index.tsx [hunk 1]
```diff
@@ -1,97 +1,170 @@
...
+import { type CoinInfo } from "@/hooks/useCoinInfo";
+import { MOCK_MARKETS } from "@/mockdata/markets";
 import { type ApiMarket } from "@/types/api";
+import { fromRawCoinAmount } from "@/utils/coin";
 import { TypeTag } from "@/utils/TypeTag";
 
-const FAUCET_ADDR =
-  "0x7c36a610d1cde8853a692c057e7bd2479ba9d5eeaeceafa24f125c23d2abf942";
+const TYPE_TAGS = [
+  TypeTag.fromString(`${FAUCET_ADDR}::test_eth::TestETHCoin`),
```
[the remark this answers is not in the document]
Comment (bogberry) [reply src/typescript/frontend/src/pages/faucet/index.tsx]: Ah yup fixed

File: src/typescript/frontend/src/pages/faucet/index.tsx [hunk 2]
```diff
@@ -1,97 +1,170 @@
...
-          }}
-          disabled={loading}
-        >
-          {loading ? "Loading..." : `Get ${coinInfo.data?.symbol}`}
-        </Button>
-      </ConnectedButton>
-    </div>
+  const balanceQueries = useQueries({
+    queries: coinInfoList.map((coinInfo, i) => ({
+      queryKey: ["balance", coinInfo.name, account?.address],
```
[the remark this answers is not in the document]
Comment (bogberry) [reply src/typescript/frontend/src/pages/faucet/index.tsx]: Good call that is cleaner and eliminates the need for this `invalidateAllBalances` function

And yes invalidating all of an account's balances should basically occur every time a user disconnects a wallet, whereas I don't ever see the need to invalidate all balances for a specific coin, so this is a better ordering

File: src/typescript/frontend/src/pages/faucet/index.tsx [hunk 3]
```diff
@@ -1,97 +1,170 @@
...
-  const faucetFunds = useCallback(async () => {
-    if (!account || !coinInfo.data) return;
-    await signAndSubmitTransaction({
-      type: "entry_function_payload",
-      function: `${FAUCET_ADDR}::test_coin::mint`,
-      type_arguments: [coinTypeTag.toString()],
-      arguments: [Math.floor(amount * 10 ** coinInfo.data.decimals)],
-    });
-  }, [account, signAndSubmitTransaction]);
+  const [isLoadingArray, setIsLoadingArray] = useState<boolean[]>(
```
[the remark this answers is not in the document]
Comment (bogberry) [reply src/typescript/frontend/src/pages/faucet/index.tsx:35]: Don't think I agree with this one - moving `loading` to the individual component level also means moving `mintCoin` there as well, and so we'd have to have `signAndSubmitTransaction` and `queryClient` there too

Having a single instance at the page level seems cleaner & simpler, and I don't think making sure we don't introduce bugs with the boolean array when we make changes in the future will be that difficult

File: src/typescript/frontend/src/pages/faucet/index.tsx [hunk 4]
```diff
@@ -1,97 +1,170 @@
...
+          ]);
+        }),
+      );
+    };
+    invalidateAllBalances();
+  }, [account?.address, coinInfoList, queryClient]);
+
+  const mintCoin = useCallback(
+    async (typeTag: TypeTag, i: number) => {
+      setIsLoadingArray((isLoadingArray) => [
```
[the remark this answers is not in the document]
Comment (bogberry) [reply src/typescript/frontend/src/pages/faucet/index.tsx:63]: yeah that should never happen, `isLoadingArray` gets initialized with the length of `TYPE_TAGS`, and every mutation preserves length

File: src/typescript/frontend/src/pages/faucet/index.tsx [hunk 5]
```diff
@@ -1,97 +1,170 @@
...
+        ...isLoadingArray.slice(0, i),
+        true,
+        ...isLoadingArray.slice(i + 1),
+      ]);
+      try {
+        await signAndSubmitTransaction({
+          type: "entry_function_payload",
+          function: `${FAUCET_ADDR}::test_coin::mint`,
+          type_arguments: [typeTag.toString()],
+          arguments: [Math.floor(AMOUNTS[i] * 10 ** coinInfoList[i].decimals)],
```
[the remark this answers is not in the document]
Comment (bogberry) [reply src/typescript/frontend/src/pages/faucet/index.tsx:73]: I don't think we need to be too paranoid about floating point errors here - it's just one multiplication by a power of 10, we can just check that there aren't any floating point errors with the amounts we've chosen, and even if there were this is just a mint function for testing so it's not like there's serious consequences

File: src/typescript/frontend/src/pages/faucet/index.tsx [hunk 6]
```diff
@@ -1,97 +1,170 @@
...
+        });
+        await queryClient.invalidateQueries([
+          "balance",
+          coinInfoList[i].name,
+          account?.address,
+        ]);
+      } catch (e) {
+        if (e instanceof Error) {
```
[the remark this answers is not in the document]
Comment (bogberry) [reply src/typescript/frontend/src/pages/faucet/index.tsx:81]: It should only ever throw an error but yeah I guess never assume there aren't bugs in other peoples code

File: src/typescript/frontend/src/pages/faucet/index.tsx [hunk 7]
```diff
@@ -1,97 +1,170 @@
...
+      </div>
+    </>
   );
 }
 
-export const getStaticProps: GetStaticProps = () => {
+export const getStaticProps: GetStaticProps = async () => {
+  const aptosClient = new AptosClient(RPC_NODE_URL);
+
+  const coinInfoList = await Promise.all(
```
[the remark this answers is not in the document]
Comment (bogberry) [reply src/typescript/frontend/src/pages/faucet/index.tsx:140]: When `getStaticProps` throws an error, the page continues to be served with the results of the last successful invocation, so propagating the error is actually the behavior we want.
<https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#error-handling-and-revalidation>

It retries when the next request is made, so I think it's best not to have any retry logic in here, as we delay serving the version of the page based on the last successful run. There are cases where serving incomplete but recent data is preferable to complete data fetched sometime in the past, but I don't think this is one of them

**Pull request merged**